jupyterhub/alembic/versions/651f5419b74d_api_token_scopes.py

```diff
@@ -13,16 +13,35 @@
 
 import sqlalchemy as sa
 from alembic import op
-from sqlalchemy import Column, ForeignKey, Table
+from sqlalchemy import Column, ForeignKey, Table, text
 from sqlalchemy.orm import relationship
 from sqlalchemy.orm.session import Session
 
-from jupyterhub import orm, roles, scopes
+from jupyterhub import orm, roles
+
+
+def access_scopes(oauth_client: orm.OAuthClient, db: Session):
+    """Return scope(s) required to access an oauth client
+    This is a clone of `scopes.access_scopes` without using
+    the `orm.Service`
+    """
+    scopes = set()
+    if oauth_client.identifier == "jupyterhub":
+        return frozenset()
+    spawner = oauth_client.spawner
+    if spawner:
+        scopes.add(f"access:servers!server={spawner.user.name}/{spawner.name}")
+    else:
+        statement = f"SELECT * FROM services WHERE oauth_client_id = '{oauth_client.identifier}'"
+        service = db.execute(text(statement)).fetchall()
```

Comment:
Rather than format strings, let's use [safe variable injection](https://docs.sqlalchemy.org/en/14/core/tutorial.html#using-textual-sql):

```suggestion
        statement =f"SELECT * FROM services WHERE oauth_client_id = :identifier"
        service = db.execute(text(statement), {"identifier": oauth_client.identifier}).fetchall()
```